### ISSUE TITLE
[Feature] Wildcard for type Id

### ISSUE THREAD

**Issue by xXAvoraXx** [openfga/openfga#803]
Hello, sometimes there may be a need to group the types into Ids as follows;

user:john
relation: access
object: permission:contoso/action/action1

user:john
relation: access
object: permission:contoso/component/component1

Is it possible to create a usage scenario like below, so I can check the nested grouped privileges;

user:john
relation: access
object: permission:contoso/action/*

user:john
relation: access
object: permission:contoso/*

user:john
relation: access
object: permission:contoso/component/*

**Comment by guipace** [comment]
I'm also interested in this use case. I'd like a group of users to have access to all objects of a type. Is this possible currently?

**Comment by aaguiarz** [comment]
@guipace 

You should be able to model that with OpenFGA currently, e.g. the example below, all admins can read all the documents from a specific organization:

`model
  schema 1.1
type user
type organization
  relations
    define admins : [user]
    
type document
  relations
    define parent: [organization]
    define can_read: admins from parent`

**Comment by aaguiarz** [comment]
@xXAvoraXx 

I think you can model that in a similar way:

```
model
  schema 1.1
type user
type organization
  relations
    define member: [user]
    define admin: [user]
    
type action
  relations
    define parent: [organization]
    define access: admin from parent
        
type component
  relations
    define parent: [organization]
    define access: admin from parent
```

If you want John to have permissions to access every component/action for the contoso organization, you'd add it to the contoso admin group.

**Comment by guipace** [comment]
@aaguiarz thanks for the quick response. I currently have a very similar model to what you proposed. I guess my question here is if there's a way to avoid having to create tuples for each individual document (or action / component in @xXAvoraXx 's case)?

If I create the following tuple as follows:
User = user:john
Relation = admin
Object = organization:my_org

I still have to create a tuple defining the organization to each document separately if I want john as an admin to have access to each document:
User = organization:my_org
Relation = parent
Object = document:1

Is there a way to wildcard the document so it applies to all documents without having to explicitly define it each time?

**Comment by aaguiarz** [comment]
If I understand correctly, in the context of the model I described above, you want to get a positive result when you call `check(user:john, can_read, document:1)` without adding a tuple for `document:1`? 

Wouldn't that be equivalent to add the permission at the organization level?

```
type organization
  relations
    define member: [user]
    define admin: [user] 
    define can_read_all_documents : admin
```

That would imply you calling check on the organization, not the document `check(user:john, can_read_all_documents, org:1)`

**Comment by guipace** [comment]
Hum, I guess that's one way of going about it. I'm new to ReBAC and OpenFGA.

Yes, your assumption is correct. The idea is that the consuming application wouldn't need to know what org a user belongs to, just perhaps their ID and OpenFGA would infer the relationship down to the document level.

**Comment by xXAvoraXx** [comment]
What I'm trying to do is actually:

All actions on the backend side correspond to a type called action in an openfga service.

I divided the frontend side into three,
route: Path extensions that represent pages.
component: Components inside the pages. Each component can have a route parent or component parent.

action: Represents the actions on the backend side.

Access relation tuples are added to the users for the relevant "route, component and action", so we have full control of what can and cannot be seen on the frontend.


Now when I give an action tuple to a user, I can check if that user has the tuple for that action. However, in order for some components to work, the user must have permission on the relevant action type. But the problem here is that it is unnecessary to allow 2 times for both the component and the action. To solve this, I made a relationship between component and action, if a user has permission on the related component type, they can access this action type.

But the problem here is that while running the check query on the backend, I check whether the user has access to the relevant action. If it does not have permission to the Action type, I have made the necessary connection to the component type, but I cannot make any progress because I am using the context structure here.

You can access the relevant schema from the link below.
https://play.fga.dev/stores/create/?id=01H16SNVJQHE8B3BMZ9FH5J405

We provide automation solutions to many businesses. After logging in, businesses manage their own businesses from a panel screen.

System administrators design a plan and define action, component, route types in the plan. It must be defined once in the system.

- Businesses subscribe to a plan.
- Users become members of the business.
- A user can be a member of more than one business.
-- I am able to resolve this last item with contextual_tuple method.

If the problem is that the action type denies can_access access, this time it needs to check the isViewer relationship, but when the component type is reached, the permission_in_context relationship is not in the tuples in the check method, so even if the user has permission in the relevant component type, access cannot be achieved.

I solve this problem by defining each action over and over for each business, but this time it both inflates the database and makes the openfga service very tiring. For these reasons, I started to develop the scheme I shared in the link above. The problems in this scheme are the reasons I explained above.


The following image will summarize the problem.
![image](https://github.com/openfga/openfga/assets/6988989/576b8ebf-598d-42d6-97f5-c8d99ab024dc)

The following image shows a flow while the check method is running.

In this example, if the user is denied access from the permission:contoso/action/action1 type, the action type goes to the component type with which it is related and looks for the user's relationship with that component. However, we cannot define it as a contextual_tuple because we do not know which component has a relationship with the action during the query.


If I could add a relationship like the one below, I think I could find out if the user has a relationship with the relevant component.
![image](https://github.com/openfga/openfga/assets/6988989/605c34de-ebe9-4aa3-878b-5e2a7ef48070)

**Comment by xXAvoraXx** [comment]
> @xXAvoraXx
> 
> I think you can model that in a similar way:
> 
> ```
> model
>   schema 1.1
> type user
> type organization
>   relations
>     define member: [user]
>     define admin: [user]
>     
> type action
>   relations
>     define parent: [organization]
>     define access: admin from parent
>         
> type component
>   relations
>     define parent: [organization]
>     define access: admin from parent
> ```
> 
> If you want John to have permissions to access every component/action for the contoso organization, you'd add it to the contoso admin group.

The main problem in this example is to save the same action types for every business over and over. If I need to delete the relevant action from the system, this time I need to find and delete this relationship network for 100 businesses. So I started developing the new schema.
Or when a new action is defined in the system, I need to define this action for all businesses.

**Comment by robyte-ctrl** [comment]
`+1`

my use case is a bit simpler: permission to create things (since there's no ID for things that weren't yet created)

here's the model
```
model
  schema 1.1

type user

type thing
  relations
    define admin: [user]
    define viewer: [user]

    define can_create: admin
```

here is what the tuple would look like
```
{ "user": "user:john", "relation": "admin", "object": "thing:*" }
```

here's some pseudocode for the check
```
check(
  user = "user:john", // check if the user `user:john`
  relation = "can_create", // has a `can_create` relation
  object = "thing:*", // with any object of type `thing` (or "with the `thing` type" also works as an explanation for me)
  authorization_id = "some-authorization-model-id"
);

Reply: true
```

or have I perhaps missed an already existing solution for the permission to create?

---

btw, any plans to fully support one-line comments in the fga language?
the playground linter complains if I add comments anywhere outside of empty lines or if I use certain symbols in the comments, like `(`, `)` or backticks (same with the vscode extension, although it doesn't complain about `(` and `)` for some reason)

**Comment by aaguiarz** [comment]
Hi @robyte-ctrl 

For that use case, in general you'd have another type that's higher in the hierarchy, e.g.

```
model
  schema 1.1

type user

type system
   relations
      define admin : [user]
      define can_create_thing : admin

type thing
  relations
    define system : [system]
    define viewer: [user]

    define can_create: admin from system
```

In B2B multi-tenant scenarios the `system` would be the `tenant` and you'll define tenant-level permissions there. In a single tenant scenario `system` would be the app itself.

Does this help?

About comments, the VS Code tool supports comments if they start at the beginning of the line. We'll improve that so they can start anywhere. We'll be soon using the same language engine for VS code and the Playground so they'll be more consistent.

**Comment by aaguiarz** [comment]
Yeah sorry, I made a mistake, `can_create` can only live in the `system` type.

The `admin` relationship might also need to be in the `system`, unless you plan to add an `admin` relationship for every `thing`. If you do that, you'd need to establish the system/thing relationship too.

**Comment by unkali** [comment]
Hello @aaguiarz , I am also facing a similar problem and would like some modelling guidance from you on the same if possible. 

Below is my current model
```
type user

type org
  relations
    define associated_user: [user]

type car
  relations
    define owner: [org]
    define is_blocked: [org]
    define shared_with: [org, org:*]
    define can_use: (owner or shared_with) but not is_blocked
   ### define can_view: can_use or admin_member?? ###
```

So, in our model we have multiple orgs that own cars and they can choose to share this car with specific orgs, ALL orgs or block specific orgs from using it. Now I have a use case where a certain admin org (eg org id = 1) gets view access to ALL cars from ALL orgs by default. I do not want to create a tuple for every car with shared_with relationship for this org. Do you have any recommendations?
